src/inewsFtp.ts

```diff
@@ -0,0 +1,447 @@
+				let promise: Promise<FtpClient>
+
+				if (reconnectAttempts > 0) {
+					promise = this.disconnect().then(async () => {
+						return connect()
+					})
+				} else {
+					promise = connect()
+				}
+
+				reconnectAttempts++
+
+				try {
+					const ftpConn = await promise
+					return ftpConn
+				} catch (error) {
+					if (
+						typeof this.config.reconnectAttempts !== 'number' ||
+						this.config.reconnectAttempts < 0 ||
+						reconnectAttempts < this.config.reconnectAttempts
+					) {
+						if (typeof this.config.reconnectTimeout === 'number' && this.config.reconnectTimeout > 0) {
+							await sleep(this.config.reconnectTimeout)
+						}
+
+						return attemptReconnect()
+					} else {
+						throw error
+					}
+				}
+			}
+
+			const connect = async (): Promise<FtpClient> => {
+				return new Promise((resolve, reject) => {
+					let returned = false
+
+					const onReady = () => {
+						if (!returned) {
+							returned = true
+							this._currentDir = null
+							removeListeners()
+							resolve(this._ftpConn)
+						}
+					}
+
+					const onError = (error: NodeJS.ErrnoException) => {
+						if (!returned) {
+							returned = true
+							removeListeners()
+							reject(error)
+						}
+					}
+
+					const onEnd = () => {
+						if (!returned) {
+							returned = true
+							removeListeners()
+							resolve(this._ftpConn)
+						}
+					}
+
+					const removeListeners = () => {
+						this._ftpConn.removeListener('ready', onReady)
+						this._ftpConn.removeListener('error', onError)
+						this._ftpConn.removeListener('end', onEnd)
+					}
+
+					this._currentHost = this.config.hosts[this._reconnectAttempts++ % this.config.hosts.length] // cycle through servers
+
+					this._setStatus('connecting')
+
+					const ftpConnConfig = {
+						host: this._currentHost,
+						user: this.config.user,
+						password: this.config.password,
+					}
+
+					this._ftpConn.once('ready', onReady)
+					this._ftpConn.once('error', onError)
+					this._ftpConn.once('end', onEnd) // workaround in case error is not emitted on timeout
+					this._ftpConn.connect(ftpConnConfig)
+				})
+			}
+
+			// Kick off the connection attempt.
+			attemptReconnect()
+				.then((ftpClient) => resolve(ftpClient))
+				.catch((error) => reject(error))
+		})
+
+		return this._connectionPromise
+	}
+
+	async disconnect(): Promise<boolean> {
+		return new Promise((resolve) => {
+			if (this._ftpConn.connected) {
+				this.once('end', () => {
+					resolve(true)
+				})
+				this._ftpConn.end()
+			} else resolve(true)
+		})
+	}
+
+	async list(directory: string): Promise<INewsFTPStoryOrQueue[]> {
+		await this.connect()
+		await this._cwd(directory)
+
+		return new Promise((resolve, reject) => {
+			this._ftpConn.list((error, list) => {
+				if (error) reject(error)
+				else {
+					const files: INewsFTPStoryOrQueue[] = []
+					if (Array.isArray(list)) {
+						;(list as Array<FtpClient.ListingElement | string>).forEach((listItem) => {
```

Comment:
Would be nice with consistent use of array type syntax. Can we change this to:
```suggestion
						;(list as (FtpClient.ListingElement | string)[]).forEach((listItem) => {
```